### PR TITLE
Make random implementations serializable

### PR DESCRIPTION
**Pull request by CommanderTvis**
Make Random and XorWowRandom expect classes and add java.io.Serializable
interface to JVM Random implementations where possible.

Related issue: https://youtrack.jetbrains.com/issue/KT-25571

### CODE REVIEW

**Comment by CommanderTvis** [comment]
The problem is that `kotlin.random.Random` objects can't be passed as Serializable or stored inside other Serializable objects.

So, this PR makes `kotlin.random.Random` and `kotlin.random.XorWowRandom` `expect` classes to provide `java.io.Serializable` interface in the JVM platform. Functions `defaultPlatformRandom` and `doubleFromParts` were removed and inlined inside the `actual` declaration of `Random`. 

JVM platform `Random` implementations were made objects and were marked interface `java.io.Serializable` too.

**Comment by ilya-g** [comment]
We'd better avoid code duplication at that scale. Could you elaborate why it is required to split so much code into expect and actual?

**Comment by CommanderTvis** [comment]
> We'd better avoid code duplication at that scale. Could you elaborate why it is required to split so much code into expect and actual?

Because only `expect/actual class` can implement an interface only on a certain platform.

**Comment by ilya-g** [comment]
Still better to find some workarounds. For example introduce an internal expect Serializable interface and implement it in common classes. In fact, there's already one in stdlib — `kotlin.io.Serializable`.

**Comment by CommanderTvis** [comment]
JVM Serializable also requires weird methods like `readResolve`. It makes no sense to have them on other platforms (even as internal or private)

**Comment by ilya-g** [comment]
Private unused methods are optimized out during tree shaking, so they won't hurt either. Still better than code duplication.

**Comment by CommanderTvis** [comment]
@ilya-g OK, I've deduplicated the classes.

File: libraries/stdlib/jvm/src/kotlin/internal/PlatformImplementations.kt [hunk 1]
```diff
@@ -51,20 +52,28 @@ internal val IMPLEMENTATIONS: PlatformImplementations = run {
     val version = getJavaVersion()
     if (version >= 0x10008) {
         try {
-            return@run castToBaseType<PlatformImplementations>(Class.forName("kotlin.internal.jdk8.JDK8PlatformImplementations").newInstance())
-        } catch (e: ClassNotFoundException) { }
+            return@run castToBaseType<PlatformImplementations>(
+                Class.forName("kotlin.internal.jdk8.JDK8PlatformImplementations").newInstance()
+            )
+        } catch (e: ClassNotFoundException) {
+        }
```
Comment (ilya-g) [inline libraries/stdlib/jvm/src/kotlin/internal/PlatformImplementations.kt:59]: Please revert formatting changes unrelated to the code regions you have changed.

File: libraries/stdlib/jdk8/src/kotlin/random/jdk8/PlatformThreadLocalRandom.kt [hunk 2]
```diff
@@ -8,15 +8,20 @@ package kotlin.random.jdk8
 import java.util.concurrent.ThreadLocalRandom
 
 @Suppress("INVISIBLE_REFERENCE", "INVISIBLE_MEMBER", "CANNOT_OVERRIDE_INVISIBLE_MEMBER")
-internal class PlatformThreadLocalRandom : kotlin.random.AbstractPlatformRandom() {
+internal object PlatformThreadLocalRandom : kotlin.random.AbstractPlatformRandom(), Serializable {
+    private const val serialVersionUID = -8182665159126504406L
```
Comment (ilya-g) [inline libraries/stdlib/jdk8/src/kotlin/random/jdk8/PlatformThreadLocalRandom.kt:12]: Why such `serialVersionUID` values, do they mean something? I believe we can start with `serialVersionUID = 1L` for all serializable entities.

File: libraries/stdlib/jdk8/src/kotlin/random/jdk8/PlatformThreadLocalRandom.kt [hunk 3]
```diff
@@ -8,15 +8,20 @@ package kotlin.random.jdk8
...
     override fun nextDouble(until: Double): Double = ThreadLocalRandom.current().nextDouble(until)
 
-//     do not delegate this, as it's buggy in JDK8+ (up to 11 at the moment of writing)
-//     override fun nextDouble(from: Double, until: Double): Double = ThreadLocalRandom.current().nextDouble(from, until)
-}
\ No newline at end of file
+    //     do not delegate this, as it's buggy in JDK8+ (up to 11 at the moment of writing)
+    //     override fun nextDouble(from: Double, until: Double): Double = ThreadLocalRandom.current().nextDouble(from, until)
+
+    private fun readResolve(): Any = PlatformThreadLocalRandom
```
Comment (ilya-g) [inline libraries/stdlib/jdk8/src/kotlin/random/jdk8/PlatformThreadLocalRandom.kt:26]: I think such internal implementations should not be serializable at all. 
`Random.Default` manages its singleton serialization entirely and therefore it shouldn't expose any implementation details in a serialized form.

File: libraries/stdlib/jvm/src/kotlin/random/PlatformRandom.kt [hunk 4]
```diff
@@ -47,17 +47,26 @@ internal abstract class AbstractPlatformRandom : Random() {
...
+
     override val impl: java.util.Random
         get() = implStorage.get()
+
+    private fun readResolve(): Any = FallbackThreadLocalRandom
 }
 
-private class PlatformRandom(override val impl: java.util.Random) : AbstractPlatformRandom()
+private class PlatformRandom(override val impl: java.util.Random) : AbstractPlatformRandom(), Serializable {
+    companion object {
```
Comment (ilya-g) [inline libraries/stdlib/jvm/src/kotlin/random/PlatformRandom.kt:66]: Companion object with just a single private const should be private.

File: libraries/stdlib/src/kotlin/random/Random.kt [hunk 5]
```diff
@@ -267,10 +267,12 @@ public abstract class Random {
      *
      * @sample samples.random.Randoms.defaultRandom
      */
-    companion object Default : Random() {
-
+    companion object Default : Random(), Serializable {
+        private const val serialVersionUID = 6449517406788167720L
         private val defaultRandom: Random = defaultPlatformRandom()
```
Comment (ilya-g) [inline libraries/stdlib/src/kotlin/random/Random.kt:272]: Better to use `writeReplace` to avoid exposing `defaultRandom` field.

File: libraries/stdlib/jdk8/src/kotlin/internal/jdk8/JDK8PlatformImplementations.kt [hunk 6]
```diff
@@ -29,13 +29,13 @@ internal open class JDK8PlatformImplementations : JDK7PlatformImplementations()
...
 
-        val range = matcher.start(name)..matcher.end(name) - 1
-        return if (range.start >= 0)
+        val range = matcher.start(name) until matcher.end(name)
+        return if (range.first >= 0) {
             MatchGroup(matcher.group(name), range)
-        else
+        } else
             null
     }
```
Comment (ilya-g) [inline libraries/stdlib/jdk8/src/kotlin/internal/jdk8/JDK8PlatformImplementations.kt:37]: This change should be in a separate commit, or even better, in a separate PR.

**Review by ilya-g — CHANGES_REQUESTED** [review]

File: libraries/stdlib/jdk8/src/kotlin/random/jdk8/PlatformThreadLocalRandom.kt (hunk 2 above)
Comment (CommanderTvis) [reply libraries/stdlib/jdk8/src/kotlin/random/jdk8/PlatformThreadLocalRandom.kt:12]: I used JDK [serialver](https://docs.oracle.com/javase/7/docs/technotes/tools/solaris/serialver.html) utility that creates something like hash, which has to be changed if the class's structure is changed

File: libraries/stdlib/src/kotlin/random/Random.kt (hunk 5 above)
Comment (CommanderTvis) [reply libraries/stdlib/src/kotlin/random/Random.kt:272]: What should `writeReplace` return then?

Comment (CommanderTvis) [reply libraries/stdlib/src/kotlin/random/Random.kt:272]: @ilya-g

Comment (ilya-g) [reply libraries/stdlib/src/kotlin/random/Random.kt:272]: Just an empty serializable class whose `readResolve` returns `Random.Default`.

File: libraries/stdlib/jdk8/src/kotlin/random/jdk8/PlatformThreadLocalRandom.kt [hunk 7]
```diff
@@ -8,7 +8,7 @@ package kotlin.random.jdk8
 import java.util.concurrent.ThreadLocalRandom
 
 @Suppress("INVISIBLE_REFERENCE", "INVISIBLE_MEMBER", "CANNOT_OVERRIDE_INVISIBLE_MEMBER")
-internal class PlatformThreadLocalRandom : kotlin.random.AbstractPlatformRandom() {
+internal object PlatformThreadLocalRandom : kotlin.random.AbstractPlatformRandom() {
```
Comment (ilya-g) [inline libraries/stdlib/jdk8/src/kotlin/random/jdk8/PlatformThreadLocalRandom.kt:11]: I don't see much sense in transforming it to `object`. It is instantiated once and then is kept in a private property of `Random` companion object. So no need to have another instance where its reference is stored.
Same for `FallbackThreadLocalRandom`

File: libraries/stdlib/src/kotlin/random/Random.kt [hunk 8]
```diff
@@ -267,10 +267,12 @@ public abstract class Random {
      *
      * @sample samples.random.Randoms.defaultRandom
      */
-    companion object Default : Random() {
-
+    companion object Default : Random(), Serializable {
+        private const val serialVersionUID = 6449517406788167720L
```
Comment (ilya-g) [inline libraries/stdlib/src/kotlin/random/Random.kt:271]: I still don't get what bad can happen if we set `serialVersionUID = 1L` in all such classes?

File: libraries/stdlib/jvm/test/random/RandomJvmSerializationTest.kt [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+import test.io.serializeAndDeserialize
+import kotlin.random.Random
+import kotlin.random.asJavaRandom
+import kotlin.random.asKotlinRandom
+import kotlin.test.Test
+import kotlin.test.assertEquals
+import kotlin.test.assertSame
+
+class RandomJvmSerializationTest {
```
Comment (ilya-g) [inline libraries/stdlib/jvm/test/random/RandomJvmSerializationTest.kt:16]: After we settle with serialized representation and serialVersionUID, let's add some serialization stability tests.
Something like https://github.com/JetBrains/kotlin/blob/1.4.20/libraries/stdlib/jvm/test/collections/CollectionJVMTest.kt#L192-L207

File: libraries/stdlib/js/src/kotlin/random/PlatformRandom.kt [hunk 10]
```diff
@@ -13,5 +13,6 @@ internal actual fun defaultPlatformRandom(): Random =
 
 private val INV_2_26: Double = 2.0.pow(-26)
 private val INV_2_53: Double = 2.0.pow(-53)
+
 internal actual fun doubleFromParts(hi26: Int, low27: Int): Double =
-    hi26 * INV_2_26 + low27 * INV_2_53
\ No newline at end of file
+    hi26 * INV_2_26 + low27 * INV_2_53
```
Comment (ilya-g) [inline libraries/stdlib/js/src/kotlin/random/PlatformRandom.kt:18]: Unrelated change

File: libraries/stdlib/src/kotlin/random/Random.kt (hunk 5 above)
Comment (ilya-g) [reply libraries/stdlib/src/kotlin/random/Random.kt:272]: On the other hand, we could rely on the fact that companion object doesn't have any fields and keeps all its state as static in the containing class, but IMO that's a too fragile assumption.

Comment (CommanderTvis) [reply libraries/stdlib/src/kotlin/random/Random.kt:272]: Then I'll make defaultRandom a property without backing field, since *ThreadLocalRandom is an `object`.

File: .idea/misc.xml [hunk 11]
```diff
@@ -1,5 +1,9 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <project version="4">
+  <component name="ASMIdeaPluginConfiguration">
+    <asm skipDebug="false" skipFrames="false" skipCode="false" expandFrames="false" />
+    <groovy codeStyle="LEGACY" />
+  </component>
```
Comment (ilya-g) [inline .idea/misc.xml:6]: Unrelated change

File: libraries/stdlib/src/kotlin/random/Random.kt [hunk 12]
```diff
@@ -267,9 +267,13 @@ public abstract class Random {
      *
      * @sample samples.random.Randoms.defaultRandom
      */
-    companion object Default : Random() {
+    companion object Default : Random(), Serializable {
+        private const val serialVersionUID = 0L
 
-        private val defaultRandom: Random = defaultPlatformRandom()
+        private val defaultRandom: Random
+            get() = defaultPlatformRandom()
```
Comment (ilya-g) [inline libraries/stdlib/src/kotlin/random/Random.kt:274]: We definitely do not want to evaluate this again on each call to Random.Default's methods. Please revert to storing the implementation here and resort to serialization surrogate approach.

File: libraries/stdlib/jvm/test/random/RandomSerializationTest.kt [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+    fun defaultIsSerializable() {
+        val instance = Random
+        assertSame(instance, serializeAndDeserialize(instance))
+    }
+
+    private fun testPersistedDeserialization(hexValue: String, expected: Any) =
+        assertEquals(expected = expected, actual = deserializeFromHex(hexValue))
+
+    @Test
+    fun deserializeDefault() = testPersistedDeserialization(
```
Comment (ilya-g) [inline libraries/stdlib/jvm/test/random/RandomSerializationTest.kt:28]: In this test we should use `assertSame` check. In the test with predictable Random, we can compare several `next*` values with the expected instance.

**Review by ilya-g — CHANGES_REQUESTED** [review]

File: libraries/stdlib/src/kotlin/random/Random.kt (hunk 12 above)
Comment (CommanderTvis) [reply libraries/stdlib/src/kotlin/random/Random.kt:274]: In JVM, this method will be 100% inlined.

File: .idea/misc.xml (hunk 11 above)
Comment (CommanderTvis) [reply .idea/misc.xml:6]: Oh, sorry.

File: libraries/stdlib/jvm/test/random/RandomSerializationTest.kt (hunk 13 above)
Comment (CommanderTvis) [reply libraries/stdlib/jvm/test/random/RandomSerializationTest.kt:28]: How many of them?

File: libraries/stdlib/src/kotlin/random/Random.kt (hunk 12 above)
Comment (CommanderTvis) [reply libraries/stdlib/src/kotlin/random/Random.kt:274]: I mean, method `kotlin/random/Random$Companion.getDefaultRandom()Lkotlin/random/Random;` will be replaced with direct access to `??????ThreadLocalRandom.INSTANCE : L???;`

File: .idea/misc.xml (hunk 11 above)
Comment (CommanderTvis) [reply .idea/misc.xml:6]: It's so cumbersome to have IDEA settings saved automatically.

File: libraries/stdlib/jvm/test/random/RandomSerializationTest.kt (hunk 13 above)
Comment (CommanderTvis) [reply libraries/stdlib/jvm/test/random/RandomSerializationTest.kt:28]: @ilya-g

File: libraries/stdlib/src/kotlin/random/Random.kt (hunk 12 above)
Comment (CommanderTvis) [reply libraries/stdlib/src/kotlin/random/Random.kt:274]: @ilya-g of course, I can discard this change, but it shouldn't make any overhead on all the platforms Kotlin supports.

File: libraries/stdlib/jvm/test/random/RandomSerializationTest.kt [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+import kotlin.random.asKotlinRandom
+import kotlin.test.Test
+import kotlin.test.assertEquals
+import kotlin.test.assertSame
+
+class RandomSerializationTest {
+    @Test
+    fun defaultIsSerializable() {
+        val instance = Random
+        assertSame(instance, serializeAndDeserialize(instance))
```
Comment (CommanderTvis) [inline libraries/stdlib/jvm/test/random/RandomSerializationTest.kt:21]: @ilya-g this test already checks that Random.default instances are same when serialized.

File: libraries/stdlib/jvm/test/random/RandomSerializationTest.kt [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+    fun deserializeDefault() = testPersistedDeserialization(
+        "ac ed 00 05 73 72 00 1c 6b 6f 74 6c 69 6e 2e 72 61 6e 64 6f 6d 2e 52 61 6e 64 6f 6d 24 44 65 66 61 75 6c 74 59 81 49 e9 14 4e a8 28 02 00 00 78 70",
+        Random
+    )
+
+    @Test
+    fun xorwowIsSerializable() {
+        val instance = Random(0)
+        val deserialized = serializeAndDeserialize(instance)
+        assertEquals(instance.nextInt(), deserialized.nextInt())
```
Comment (CommanderTvis) [inline libraries/stdlib/jvm/test/random/RandomSerializationTest.kt:37]: @ilya-g of course, we can check more than first 4 bytes the deserialized random produces.

File: libraries/stdlib/src/kotlin/random/Random.kt (hunk 12 above)
Comment (ilya-g) [reply libraries/stdlib/src/kotlin/random/Random.kt:274]: Better not to speculate about platform optimization capabilities, especially considering that this code can run on other VMs such as Android runtime, when we can write it in a way that make it sure happen without any optimization.
In other Kotlin platforms, the changes caused by this one are also totally unnecessary.

File: libraries/stdlib/jvm/test/random/RandomSerializationTest.kt (hunk 13 above)
Comment (ilya-g) [reply libraries/stdlib/jvm/test/random/RandomSerializationTest.kt:28]: I believe one call for each of the following methods would be enough: `nextInt()`, `nextLong()`, `nextDouble()`, `nextBytes()`

File: libraries/stdlib/jvm/test/random/RandomSerializationTest.kt (hunk 14 above)
Comment (ilya-g) [reply libraries/stdlib/jvm/test/random/RandomSerializationTest.kt:21]: It doesn't catch the situation when we change the default instance in an incompatible way.

**Review by ilya-g — CHANGES_REQUESTED** [review]

File: libraries/stdlib/jvm/test/random/RandomSerializationTest.kt (hunk 13 above)
Comment (CommanderTvis) [reply libraries/stdlib/jvm/test/random/RandomSerializationTest.kt:28]: Clear

File: libraries/stdlib/jvm/test/random/RandomSerializationTest.kt (hunk 14 above)
Comment (CommanderTvis) [reply libraries/stdlib/jvm/test/random/RandomSerializationTest.kt:22]: In the current state of changes, `default` has no serialized fields and has constant `serialVersionUID`. Instances will be persistent until we add non-transient field or change `serialVersinoUID`. I'm not sure how to test such a situation well.

File: libraries/stdlib/src/kotlin/random/Random.kt [hunk 16]
```diff
@@ -290,7 +295,11 @@ public abstract class Random {
 
         override fun nextBytes(array: ByteArray): ByteArray = defaultRandom.nextBytes(array)
         override fun nextBytes(size: Int): ByteArray = defaultRandom.nextBytes(size)
-        override fun nextBytes(array: ByteArray, fromIndex: Int, toIndex: Int): ByteArray = defaultRandom.nextBytes(array, fromIndex, toIndex)
+        override fun nextBytes(array: ByteArray, fromIndex: Int, toIndex: Int): ByteArray = defaultRandom.nextBytes(
+            array,
+            fromIndex,
+            toIndex
+        )
```
Comment (ilya-g) [inline libraries/stdlib/src/kotlin/random/Random.kt:302]: This change is not related. Better to keep it consistent with the other function definitions.

File: libraries/stdlib/src/kotlin/random/Random.kt [hunk 17]
```diff
@@ -267,10 +268,14 @@ public abstract class Random {
      *
      * @sample samples.random.Randoms.defaultRandom
      */
-    companion object Default : Random() {
+    companion object Default : Random(), Serializable {
+        private const val serialVersionUID = 0L
 
+        @Transient
```
Comment (ilya-g) [inline libraries/stdlib/src/kotlin/random/Random.kt:274]: Ok, I think the variant with all transient fields and `readResolve` will do (without introducing a serialization surrogate).

**Review by ilya-g — COMMENTED** [review]
Please rebase your branch on the latest `master` and ensure there are no merge commits in it.

File: libraries/stdlib/src/kotlin/random/Random.kt [hunk 18]
```diff
@@ -267,10 +268,14 @@ public abstract class Random {
      *
      * @sample samples.random.Randoms.defaultRandom
      */
-    companion object Default : Random() {
+    companion object Default : Random(), Serializable {
+        private const val serialVersionUID = 0L
```
Comment (ilya-g) [inline libraries/stdlib/src/kotlin/random/Random.kt:272]: Bad news: this const doesn't work as we intended it to. The reason is that it is compiled as a static field to the `Random` class, and not to the `Random.Default`.
So changing its value doesn't fail the test with persistent deserialization, and on the opposite, adding a public method to `Random.Default` fails the test, which indicates that `serialVersionUID` for this class is still automatic.

Given the above I think we have no other way than to use a serialization surrogate.

**Review by ilya-g — CHANGES_REQUESTED** [review]

File: libraries/stdlib/src/kotlin/random/Random.kt (hunk 18 above)
Comment (CommanderTvis) [reply libraries/stdlib/src/kotlin/random/Random.kt:272]: What do you mean by serialization surrogate?

File: libraries/stdlib/jvm/test/random/RandomSerializationTest.kt [hunk 19]
```diff
@@ -0,0 +1,72 @@
...
+        assertEquals(first.nextInt(), second.nextInt())
+        assertEquals(first.nextDouble(), second.nextDouble())
+        assertEquals(first.nextLong(), second.nextLong())
+        assertTrue(first.nextBytes(64).contentEquals(second.nextBytes(64)))
+    }
+
+    @Test
+    fun deserializeDefault() = assertSame(
+        expected = Random,
+        actual = deserializeFromHex("ac ed 00 05 73 72 00 1c 6b 6f 74 6c 69 6e 2e 72 61 6e 64 6f 6d 2e 52 61 6e 64 6f 6d 24 44 65 66 61 75 6c 74 59 81 49 e9 14 4e a8 28 02 00 00 78 70")
```
Comment (ilya-g) [inline libraries/stdlib/jvm/test/random/RandomSerializationTest.kt:43]: Please also add a similar test for XorwowRandom.

File: libraries/stdlib/src/kotlin/random/Random.kt (hunk 18 above)
Comment (CommanderTvis) [reply libraries/stdlib/src/kotlin/random/Random.kt:272]: Clear

File: libraries/stdlib/jvm/test/random/RandomSerializationTest.kt (hunk 19 above)
Comment (CommanderTvis) [reply libraries/stdlib/jvm/test/random/RandomSerializationTest.kt:43]: OK

File: libraries/stdlib/src/kotlin/random/Random.kt (hunk 18 above)
Comment (ilya-g) [reply libraries/stdlib/src/kotlin/random/Random.kt:272]: Can be a nested class in `Random` or `Random.Default`

File: libraries/stdlib/jvm/test/random/RandomSerializationTest.kt (hunk 19 above)
Comment (CommanderTvis) [reply libraries/stdlib/jvm/test/random/RandomSerializationTest.kt:43]: It is impossible since serialization produces a new instance with state restored from bytes, so instances of XorWow will be neither same nor equal (since XorWow is not a data class)

Comment (ilya-g) [reply libraries/stdlib/jvm/test/random/RandomSerializationTest.kt:43]: Indeed, that's why I said "a similar", not "the same" :)
You can use `testRandomsHaveSameState` to check that the state has been restored correctly.

Comment (CommanderTvis) [reply libraries/stdlib/jvm/test/random/RandomSerializationTest.kt]: OK

**Review by ilya-g — APPROVED** [review]

**Comment by CommanderTvis** [comment]
@ilya-g how to run API dump?

**Comment by ilya-g** [comment]
To perform API dump, you can use two shared run configurations in Kotlin project:
- `Test: public API binary compatibility validator, overwrite results` (dumps JVM public binary API)
- `Test: stdlib-js public kotlin api test, overwrite results` (dumps JS source API)

**Comment by ilya-g** [comment]
I've merged it. Thank you for the good work!
